### ISSUE TITLE
sum post-op

### ISSUE THREAD

**Issue by dev2y** [intel/mkl-dnn#392]
Hi, I am trying to add sum post_op to BatchNorm and following is my code. I assume this will take batch normalization output as one of the inputs to sum post_op. But this sum has two inputs. I am wondering how to make this sum to use second input (from another branch) 

Can you please give me a sample code or point me to example? I couldn't find it in mkldnn gtests or in examples.

```
      mkldnn::primitive_attr attr;
      attr.set_int_output_round_mode(mkldnn::round_mode::round_nearest);
      const float ops_scale = 1.f;
      mkldnn::post_ops ops;
      ops.append_sum(ops_scale);
      attr.set_post_ops(ops);

      BatchNormFwd_pd.reset(new batch_normalization_forward::primitive_desc(BatchNormFwd_desc, attr,
                ep.getMKLDNNSubEngine(subEngineIndex)));
```

**Comment by igorsafo** [comment]
Hi @dev2y,
BatchNorm doesn't support sum post_op.  In general when sum post op is used second branch provided as a destination for a primitive: dst <- sum( primitive(src, ...), dst*scale )

**Comment by dev2y** [comment]
@ igorsafo 
I've the code from mkldnn examples. Can you please help me how I can add second (or third) input for sum post op?

Thanks

    auto conv1_src_md = memory::desc(
            { conv1_src_tz }, memory::data_type::f32, memory::format::any);
    auto conv1_bias_md = memory::desc(
            { conv1_bias_tz }, memory::data_type::f32, memory::format::any);
    auto conv1_weights_md = memory::desc(
            { conv1_weights_tz }, memory::data_type::f32, memory::format::any);
    auto conv1_dst_md = memory::desc(
            { conv1_dst_tz }, memory::data_type::f32, memory::format::any);

	mkldnn::primitive_attr attr;
    attr.set_int_output_round_mode(mkldnn::round_mode::round_nearest);
    const float ops_scale = 1.f;
    mkldnn::post_ops ops;
    ops.append_sum(ops_scale);
    attr.set_post_ops(ops);

    /* create a convolution */
    auto conv1_desc = convolution_forward::desc(
            prop_kind::forward_inference, convolution_direct, conv1_src_md,
            conv1_weights_md, conv1_bias_md, conv1_dst_md, conv1_strides,
            conv1_padding, conv1_padding, padding_kind::zero);
    auto conv1_prim_desc
            = convolution_forward::primitive_desc(conv1_desc, attr, cpu_engine);

    /* create reorders for data and weights if layout requested by
     * convolution is different from NCHW/OIHW */
    auto conv1_src_memory = user_src_memory;
    if (memory::primitive_desc(conv1_prim_desc.src_primitive_desc())
            != user_src_memory.get_primitive_desc()) {
        conv1_src_memory = memory(conv1_prim_desc.src_primitive_desc());
        net.push_back(reorder(user_src_memory, conv1_src_memory));
    }

**Comment by rsdubtso** [comment]
For the sum post-op, the second argument is always the destination: dst = dst + result.

**Comment by dev2y** [comment]
Sorry. I am still confused. which second argument? ops.append_sum(ops_scale); takes only one parameter.

**Comment by igorsafo** [comment]
@dev2y ,
`conv1_dst_memory.set_data_handle((void *)second_operand_buffer);`
convolution primitive will **add** conv op result to the `second_operand_buffer` buffer (instead of writing into this buffer).

**Comment by dev2y** [comment]
Thanks @igorsafo. I think I got it. But where will write the final result after sum?

**Comment by igorsafo** [comment]
In my example final result will be stored into `second_operand_buffer`:
second_operand_buffer <- conv1(src, weights) + second_operand_buffer

**Comment by dev2y** [comment]
Got it. Thank you. I will close this.
